KOGITO-2024, KOGITO-2068, KOGITO-2110 - Add human task and rules (drl) support for Serverless Workflow

tsurdilo opened the pull request:
This also allows user tasks to be placed inside embedded subprocess (tested by the serverless workflow tests)

cristianonicolai commented on jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/util/ServerlessWorkflowUtils.java:
```diff
@@ -155,16 +156,22 @@ public static String scriptFunctionScript(String script) {
 
         if (path.startsWith("$.")) {
             path = path.substring(2);
+        } else {
```
```suggestion
        } else if {
```

tsurdilo replied:
updated

cristianonicolai commented on jbpm/jbpm-serverless-workflow/src/test/java/org/jbpm/serverless/workflow/WorkflowFactoryTest.java:
```diff
@@ -232,4 +232,73 @@ public void testJoinNode() {
+        RuleFlowProcess process = new RuleFlowProcess();
+
+        Function function = new Function().withName("testfunction1").withMetadata(
+                new HashMap() {{
```
My personal preference here would be to simply create a new HashMap and assign a variable, as this style endups creating an inner class.

tsurdilo replied:
ok. updated

tsurdilo commented:
@cristianonicolai It's a little different, with serverless workflow you can define functions which execute services. A function can define a type, and we use that to map to custom impls. 
So for example a "script" type maps to a script task, "service' type maps to workitems, and "decision" type maps to a human decision. Services like dmn , other workflows, other bpmn2 process and rules all fall under "services" which can be invoked either via things like rest workitem or a custom service invocation, so they would all fall under the "service" mapping which is already supported. Hope this helps.

cristianonicolai commented:
@tsurdilo thanks for the details, I'm not following the spec in details so just thought that a decision could be either automated or done by a person.

tsurdilo commented:
@cristianonicolai I added also KOGITO-2068 to this pr as there are no reviews and it was the next thing I was going to do anyways, and it affecting the same code as the original pr.

Also fixed issue with rule nodes being placed inside composite context nodes. this is tested by the serverless workflow tests in kogito-codegen.

tsurdilo commented:
@MarianMacik @tiagodolphine Hi guys, would you mind taking a look when you have a chance please ?

tiagodolphine commented on jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/UserTaskModelMetaData.java:
```diff
@@ -160,7 +162,10 @@ private CompilationUnit compilationUnitInput() {
 
             if (variable == null) {
-                throw new IllegalStateException("Task " + humanTaskNode.getName() +" (input) " + entry.getKey() + " reference not existing variable " + entry.getValue());
+                variable = processVariableScope.findVariable(entry.getValue());
```
what about using 
Variable variable = Optional.ofNullable(variableScope.findVariable(entry.getValue()))
                    .orElse(processVariableScope.findVariable(entry.getValue()));

tsurdilo replied:
updated

tiagodolphine commented on jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/canonical/UserTaskModelMetaData.java:
```diff
@@ -251,15 +256,18 @@ private CompilationUnit compilationUnitOutput() {
-                    variable.setType(new ObjectDataType(dataOutputs.get(entry.getKey())));
-                } else {
-                    throw new IllegalStateException("Task " + humanTaskNode.getName() +" (output) " + entry.getKey() + " reference not existing variable " + entry.getValue());
+                variable = processVariableScope.findVariable(entry.getValue());
```
what about using
Variable variable = Optional.ofNullable(variableScope.findVariable(entry.getValue()))
.orElse(processVariableScope.findVariable(entry.getValue()));

tsurdilo replied:
updated

tsurdilo commented:
@tiagodolphine thanks fore review! Updated per your comments.

tsurdilo commented:
@cristianonicolai can we just merge this? idk what the qe holdup might be about....
